File: node/src/runner.ts
/** Script-type runner: fixed executable + argv array + stdin envelope
 * (port of hub/runner.py).
 *
 * Platform specifics handled here:
 *   - timeout kills the whole tree: POSIX via detached process group +
 *     kill(-pid, SIGKILL) (deliberate improvement #2 in docs/NODE-PLAN.md —
 *     the Python version only kills the direct child on POSIX), Windows via
 *     ``taskkill /T /F``
 *   - stdout/stderr are capped while streaming so a runaway process cannot
 *     exhaust memory
 *   - child gets a minimal environment, not process.env
 */

import { type ChildProcess, spawn, spawnSync } from "node:child_process";
import { accessSync, constants } from "node:fs";
import * as path from "node:path";
import { performance } from "node:perf_hooks";

import { buildRequest, parseEnvelope, type Envelope } from "./envelope.js";
import { SkillExecutionError, SkillInputError } from "./errors.js";
import {
  DEFAULT_MAX_STDOUT,
  DEFAULT_TIMEOUT,
  type SkillConfig,
  type SkillRegistry,
} from "./registry.js";
import { PathGuard, buildEnv, scrub } from "./security.js";

/** Turn a registry executable name into an argv prefix the OS accepts.
 *
 * npm-style shims (``claude.CMD`` etc.) are not directly launchable on
 * Windows and go through ``cmd /c`` with their resolved path. Extension-less
 * POSIX-style shims fall back to the bare name and the OS PATH search.
 */
export function resolveExecutable(name: string): string[] {
  if (name.includes("/") || name.includes("\\")) return [name];
  const found = whichSync(name);
  if (!found) return [name];
  const lowered = found.toLowerCase();
  if (lowered.endsWith(".cmd") || lowered.endsWith(".bat")) {
    const comspec = process.env.COMSPEC ?? "cmd.exe";
    return [comspec, "/d", "/c", found];
  }
  if (lowered.endsWith(".exe") || lowered.endsWith(".com")) return [found];
  return [name];
}

function whichSync(name: string): string | null {
  const pathEnv = process.env.PATH ?? "";
  const exts =
    process.platform === "win32"
      ? (process.env.PATHEXT ?? ".COM;.EXE;.BAT;.CMD").split(";").filter(Boolean)
      : [""];
  for (const dir of pathEnv.split(path.delimiter)) {
    if (!dir) continue;
    for (const ext of exts) {
      const candidate = path.join(dir, name + ext);
      try {
        accessSync(candidate, constants.X_OK);
        return candidate;
      } catch {
        // keep looking
      }
    }
  }
  return null;
}

function killTree(child: ChildProcess): void {
  /** Terminate the whole child tree; plain kill() leaks grandchildren. */
  if (child.pid === undefined) return;
  if (process.platform === "win32") {
    try {
      spawnSync("taskkill", ["/T", "/F", "/PID", String(child.pid)], {
        timeout: 10_000,
      });
    } catch {
      // best effort
    }
  } else {
    try {
      process.kill(-child.pid, "SIGKILL");
    } catch {
      // already gone
    }
  }
  try {
    child.kill("SIGKILL");
  } catch {
    // already gone
  }
}

/** Streaming collector that stores at most cap+1 bytes and keeps draining. */
function cappedCollector(cap: number) {
  let stored = 0;
  let total = 0;
  const chunks: Buffer[] = [];
  return {
    push(chunk: Buffer): void {
      total += chunk.length;
      if (stored <= cap) {
        const take = chunk.subarray(0, cap + 1 - stored);
        chunks.push(take);
        stored += take.length;
      }
    },
    text(): { text: string; truncated: boolean } {
      return {
        text: Buffer.concat(chunks).toString("utf8"),
        truncated: total > cap,
      };
    },
  };
}

export interface RunOptions {
  dryRun?: boolean;
  client?: string;
}

export class ScriptRunner {
  readonly guard: PathGuard;

  constructor(
    readonly registry: SkillRegistry,
    workspaceRoot: string,
  ) {
    this.guard = new PathGuard(workspaceRoot);
  }

  async run(
    skillId: string,
    inputs: Record<string, unknown>,
    { dryRun = true, client = "" }: RunOptions = {},
  ): Promise<Envelope> {
    const runtime = this.registry.runtime(skillId);
    const permissions = this.registry.permissions(skillId);
    const timeoutSeconds = Number(runtime.timeout_seconds ?? DEFAULT_TIMEOUT);
    const maxStdout = Number(runtime.max_stdout_bytes ?? DEFAULT_MAX_STDOUT);

    const confined = this.confinePaths(skillId, inputs);
    const argv = [
      ...resolveExecutable(String(runtime.executable)),
      ...((runtime.args as string[]) ?? []),
    ];
    const requestBody = buildRequest(skillId, confined, { dryRun, client });

    const workdir = this.workingDirectory(runtime);
    const env = buildEnv(
      (permissions.environment?.allow as string[] | undefined) ?? [],
    );

    const started = performance.now();
    const stdoutCap = cappedCollector(maxStdout);
    const stderrCap = cappedCollector(20_000);
    let timedOut = false;
    let exitCode: number | null = null;
    let exitSignal: NodeJS.Signals | null = null;

    const child = spawn(argv[0]!, argv.slice(1), {
      cwd: workdir,
      env,
      stdio: ["pipe", "pipe", "pipe"],
      // new process group on POSIX so the timeout can kill the whole tree
      detached: process.platform !== "win32",
    });

    // the skill may exit before consuming stdin; EPIPE then is not an error
    child.stdin?.on("error", () => {});
    child.stdin?.end(requestBody, "utf8");
    child.stdout?.on("data", (chunk: Buffer) => stdoutCap.push(chunk));
    child.stderr?.on("data", (chunk: Buffer) => stderrCap.push(chunk));

    await new Promise<void>((resolve, reject) => {
      let settled = false;
      const finish = (err?: Error) => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        err ? reject(err) : resolve();
      };
      const timer = setTimeout(() => {
        timedOut = true;
        killTree(child);
      }, timeoutSeconds * 1000);
      child.on("error", (err: NodeJS.ErrnoException) => {
        if (err.code === "ENOENT") {
          finish(
            new SkillExecutionError(
              `Executable for skill ${JSON.stringify(skillId)} not found: ${argv[0]} (${err.message})`,
            ),
          );
        } else {
          finish(
            new SkillExecutionError(
              `Skill ${JSON.stringify(skillId)} failed to start: ${err.message}`,
            ),
          );
        }
      });
      child.on("close", (code, signal) => {
        exitCode = code;
        exitSignal = signal;
        finish();
      });
    });

    if (timedOut) {
      throw new SkillExecutionError(
        `Skill ${JSON.stringify(skillId)} timed out after ${timeoutSeconds}s and was killed`,
      );
    }

    if (exitCode !== 0) {
      const stderrText = stderrCap.text().text;
      const codeLabel = exitCode ?? `signal ${exitSignal}`;
      throw new SkillExecutionError(
        `Skill ${JSON.stringify(skillId)} exited with code ${codeLabel}: ${scrub(stderrText)}`,
      );
    }

    const { text: stdoutText, truncated } = stdoutCap.text();
    const envelope = parseEnvelope(stdoutText);
    if (truncated) {
      envelope.warnings.push("stdout exceeded cap and was truncated");
    }
    envelope.duration_ms = Math.round(performance.now() - started);
    return envelope;
  }

  /** Resolve x-path-scope inputs and replace them with confined absolute paths.
   *
   * The skill never sees the caller's raw path string, only a path the hub
   * has already proven to be inside the declared permission roots.
   */
  confinePaths(
    skillId: string,
    inputs: Record<string, unknown>,
  ): Record<string, unknown> {
    const scopes = this.registry.pathScopes(skillId);
    const fs = this.registry.permissions(skillId).filesystem ?? {};
    const confined = { ...inputs };
    for (const [key, scope] of Object.entries(scopes)) {
      if (!(key in confined)) continue;
      const roots = (fs[scope] as string[] | undefined) ?? [];
      if (!roots.length) {
        throw new SkillExecutionError(
          `Input ${JSON.stringify(key)} declares x-path-scope=${JSON.stringify(scope)} but the ` +
            "registry lists no matching filesystem roots",
        );
      }
      confined[key] = this.guard.resolve(confined[key], {
        allowed: roots,
        mustExist: scope === "read",
      });
    }
    return confined;
  }

  private workingDirectory(runtime: SkillConfig): string {
    const configured = runtime.working_directory as string | undefined;
    if (!configured) return this.registry.projectRoot;
    const resolved = this.registry.resolveProject(configured);
    const projectRoot = this.registry.projectRoot;
    if (
      resolved !== projectRoot &&
      !resolved.startsWith(projectRoot + path.sep)
    ) {
      throw new SkillExecutionError(
        `working_directory ${JSON.stringify(configured)} is outside the project root`,
      );
    }
    return resolved;
  }
}

// re-export for callers that only care about input errors
export { SkillInputError };

File: hub/runner.py
"""Script-type runner: fixed executable + argv array + stdin envelope.

Windows specifics handled here:
  - timeout kills the whole tree via ``taskkill /T /F`` (``process.kill()``
    only terminates the direct child and orphans grandchildren)
  - stdout/stderr go to temp files so a runaway process cannot exhaust
    memory; caps are applied when reading back
  - child gets a minimal environment, not ``os.environ``
"""

from __future__ import annotations

import asyncio
import os
import shutil
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Any

from .envelope import build_request, parse_envelope
from .errors import SkillExecutionError
from .registry import DEFAULT_MAX_STDOUT, DEFAULT_TIMEOUT, SkillRegistry
from .security import PathGuard, build_env, scrub


def resolve_executable(name: str) -> list[str]:
    """Turn a registry executable name into an argv prefix that CreateProcess accepts.

    npm-style shims (``claude.CMD`` etc.) are not directly launchable on
    Windows and go through ``cmd /c`` with their resolved path. Extension-less
    POSIX-style shims (``~/bin/python``) are equally unlaunchable but have no
    cmd equivalent, so we fall back to the bare name and let CreateProcess do
    its native PATH search.
    """
    if os.sep in name or (os.altsep and os.altsep in name):
        return [name]
    found = shutil.which(name)
    if found is None:
        return [name]
    lowered = found.lower()
    if lowered.endswith((".cmd", ".bat")):
        comspec = os.environ.get("COMSPEC", "cmd.exe")
        return [comspec, "/d", "/c", found]
    if lowered.endswith((".exe", ".com")):
        return [found]
    return [name]


class ScriptRunner:
    def __init__(self, registry: SkillRegistry, workspace_root: str | Path):
        self.registry = registry
        self.guard = PathGuard(workspace_root)

    async def run(
        self,
        skill_id: str,
        inputs: dict[str, Any],
        *,
        dry_run: bool = True,
        client: str = "",
    ) -> dict[str, Any]:
        runtime = self.registry.runtime(skill_id)
        permissions = self.registry.permissions(skill_id)
        timeout = runtime.get("timeout_seconds", DEFAULT_TIMEOUT)
        max_stdout = runtime.get("max_stdout_bytes", DEFAULT_MAX_STDOUT)

        inputs = self.confine_paths(skill_id, inputs)
        argv = [*resolve_executable(runtime["executable"]), *runtime.get("args", [])]
        request_body = build_request(skill_id, inputs, dry_run=dry_run, client=client)

        workdir = self._working_directory(runtime)
        env = build_env(
            permissions.get("environment", {}).get("allow", []),
        )

        started = time.monotonic()
        with tempfile.TemporaryFile() as out_file, tempfile.TemporaryFile() as err_file:
            try:
                proc = await asyncio.create_subprocess_exec(
                    *argv,
                    cwd=str(workdir),
                    env=env,
                    stdin=asyncio.subprocess.PIPE,
                    stdout=out_file,
                    stderr=err_file,
                )
            except FileNotFoundError as exc:
                raise SkillExecutionError(
                    f"Executable for skill {skill_id!r} not found: {argv[0]} ({exc})"
                ) from exc
            try:
                await asyncio.wait_for(
                    proc.communicate(request_body.encode("utf-8")), timeout=timeout
                )
            except asyncio.TimeoutError:
                self._kill_tree(proc)
                raise SkillExecutionError(
                    f"Skill {skill_id!r} timed out after {timeout}s and was killed"
                )

            if proc.returncode != 0:
                stderr_text = self._read_capped(err_file, 20_000)[0]
                raise SkillExecutionError(
                    f"Skill {skill_id!r} exited with code {proc.returncode}: "
                    f"{scrub(stderr_text)}"
                )

            stdout_text, truncated = self._read_capped(out_file, max_stdout)

        envelope = parse_envelope(stdout_text)
        if truncated:
            envelope["warnings"].append("stdout exceeded cap and was truncated")
        envelope["duration_ms"] = int((time.monotonic() - started) * 1000)
        return envelope

    def confine_paths(self, skill_id: str, inputs: dict[str, Any]) -> dict[str, Any]:
        """Resolve x-path-scope inputs and replace them with confined absolute paths.

        The skill never sees the caller's raw path string, only a path the hub
        has already proven to be inside the declared permission roots.
        """
        scopes = self.registry.path_scopes(skill_id)
        fs = self.registry.permissions(skill_id).get("filesystem", {})
        confined = dict(inputs)
        for key, scope in scopes.items():
            if key not in confined:
                continue
            roots = fs.get(scope, [])
            if not roots:
                raise SkillExecutionError(
                    f"Input {key!r} declares x-path-scope={scope!r} but the "
                    "registry lists no matching filesystem roots"
                )
            resolved = self.guard.resolve(
                str(confined[key]),
                allowed=roots,
                must_exist=(scope == "read"),
            )
            confined[key] = str(resolved)
        return confined

    def _working_directory(self, runtime: dict[str, Any]) -> Path:
        configured = runtime.get("working_directory")
        if not configured:
            return self.registry.project_root
        resolved = self.registry.resolve_project(configured)
        project_root = self.registry.project_root
        if (
            resolved != project_root
            and project_root not in resolved.parents
        ):
            raise SkillExecutionError(
                f"working_directory {configured!r} is outside the project root"
            )
        return resolved

    @staticmethod
    def _read_capped(fileobj, cap: int) -> tuple[str, bool]:
        fileobj.seek(0)
        data = fileobj.read(cap + 1)
        return data.decode("utf-8", errors="replace"), len(data) > cap

    @staticmethod
    def _kill_tree(proc: asyncio.subprocess.Process) -> None:
        """Terminate the whole child tree; plain kill() leaks grandchildren."""
        if proc.returncode is not None:
            return
        try:
            subprocess.run(
                ["taskkill", "/T", "/F", "/PID", str(proc.pid)],
                capture_output=True,
                timeout=10,
            )
        except (OSError, subprocess.TimeoutExpired):
            pass
        try:
            proc.kill()
        except ProcessLookupError:
            pass
